devdoc/waiter.md

```diff
@@ -0,0 +1,231 @@
+
+### cancel_register_notification
+```
+static void cancel_register_notification(void* context)
+```
+
+`cancel_register_notification` is the cancel function given to `async_op_create` when `waiter_register_notification` is called.
+
+**SRS_WAITER_43_046: [** If the current state of the `waiter` is the same as the initial state of the `waiter` when `waiter_register_notification` was called, `cancel_register_notification` shall: **]**
+
+ **SRS_WAITER_43_047: [** increment the `EPOCH` and unset the `REGISTERED` bit by calling `interlocked_compare_exchange`. **]**
+
+ **SRS_WAITER_43_048: [** call the `notification_callback` given to `waiter_register_notification` with `context` as the `context` given to `waiter_register_notification`, `data` as `NULL` and `result` as `WAITER_RESULT_CANCELLED`. **]**
+
+
+### waiter_notify
+```c
+    MOCKABLE_FUNCTION(, WAITER_RESULT, waiter_notify, THANDLE(WAITER), waiter, THANDLE(RC_PTR), data, NOTIFY_COMPLETE_CALLBACK, notify_complete_callback, void*, context, THANDLE(ASYNC_OP)*, out_op);
+```
+
+`waiter_notify` notifies the waiter that there is data available and registers the given `notify_complete_callback` to be called when the given `data` has been consumed.
+
+**SRS_WAITER_43_024: [** If `waiter` is `NULL`, `waiter_notify` shall fail and return `WAITER_RESULT_ERROR`. **]**
+
+**SRS_WAITER_43_025: [** If `notify_complete_callback` is `NULL`, `waiter_notify` shall fail and return `WAITER_RESULT_ERROR`. **]**
+
+**SRS_WAITER_43_026: [** If `out_op` is `NULL`, `waiter_notify` shall fail and return `WAITER_RESULT_ERROR`. **]**
+
+**SRS_WAITER_43_027: [** `waiter_notify` shall call `interlocked_or` to set the `NOTIFIED` bit and obtain the current state of the `waiter`. **]**
+
+**SRS_WAITER_43_028: [** If the `NOTIFIED` bit was already set in the current state, `waiter_notify` shall fail and return `WAITER_RESULT_REFUSED`. **]**
+
+**SRS_WAITER_43_029: [** If the `REGISTERED` bit is set, `waiter_notify` shall: **]**
+
+- **SRS_WAITER_43_043: [** call `interlocked_exchange` to increment the `EPOCH` and unset the `NOTIFIED` bit and `REGISTERED` bit. **]**
+
+- **SRS_WAITER_43_030: [** call the given `notification_callback` with `context` as the `context` given to `waiter_register_notification`, `data` as the given `data` and `result` as `WAITER_RESULT_OK`. **]**
+
+- **SRS_WAITER_43_031: [** call the given `notify_complete_callback` with `context` as the given `context` and `result` as `WAITER_RESULT_OK`. **]**
+
+- **SRS_WAITER_43_032: [** return `WAITER_RESULT_SYNC`. **]**
+
+**SRS_WAITER_43_033: [** If the `REGISTERED` bit was not already set, `waiter_notify` shall: **]**
+
+- **SRS_WAITER_43_034: [** call `async_op_create` with `cancel_notify` as the cancel function. **]**
+
+- **SRS_WAITER_43_035: [** store the given `notify_complete_callback` and the given `context` in the `waiter`. **]**
+
+- **SRS_WAITER_43_036: [** store the given `data` in the `waiter` by calling `THANDLE_INITIALIZE`. **]**
```

Comment:
>store the given `data` in the `waiter` [](http://example.com/codeflow?start=27&length=38)

That's a bit weird to store this data in the waiter, as... I would see the data more related to the async_op more than the module itself. Interesting.

Reply:
The `waiter` needs to be able to access the data to complete the operation. `waiter` cannot reference the async_op because the async_op references the `waiter` (otherwise there will be a circular reference and memory leak). So the data must be stored in the `waiter`.